node: devicemgr: Graduate Kubelet DeviceManager to GA

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature


#### What this PR does / why we need it: 
Captures changes to graduate Kubelet Device Manager to GA.
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Issue: https://github.com/kubernetes/enhancements/issues/3573
#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Graduate Kubelet Device Manager to GA.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: https://github.com/kubernetes/enhancements/pull/3574
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/pull/3574
```

swatisehgal committed:
node: devicemgr: Graduate Kubelet DeviceManager to GA

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

fromanirh commented:
/sig node

fromanirh commented:
/priority important-soon
because sig-node would like to promote device manager GA in 1.26

fromanirh commented:
/triage accepted

liggitt commented:
/uncc
cc @kubernetes/sig-node-pr-reviews

ffromani reviewed (commented):
LGTM (soft, non-binding) because: this change is correct; I'm not sure the rest of the code (and the tests) should be changed to consider the feature always on. Should this be changed when we lock the feature gate, or should we do this when we remove the feature gate? Or are both acceptable?

swatisehgal committed:
node: devicemgr: GA graduation implies Feature Gate is ON by default

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: devicemgr: Remove `devicePluginEnabled` field from container mgr

With graduation of device plugins to GA in 1.26, the feature gate is
enabled by default so `devicePluginEnabled` field no longer needs to
be passed at the time of Container Manager creation.

In addition to that, we remove the `ManagerStub` as it is no longer
needed.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

fromanirh commented:
it seems the GCE script can be simplified further
```

In ./cluster/gce/config-test.sh line 160:
FEATURE_GATES=${KUBE_FEATURE_GATES:-}
^-----------^ SC2034: FEATURE_GATES appears unused. Verify use (or export if used externally).
In ./cluster/gce/config-default.sh line 259:
FEATURE_GATES="${KUBE_FEATURE_GATES:-}"
^-----------^ SC2034: FEATURE_GATES appears unused. Verify use (or export if used externally).
For more information:
  https://www.shellcheck.net/wiki/SC2034 -- FEATURE_GATES appears unused. Ver... 
```

fromanirh commented:
thanks for the updates! Removing the if guards when we lock the feature gate makes sense to me, and according to the (admittedly few) examples I managed to find, this seems the standard practice when graduating to GA.

bart0sh commented on pkg/kubelet/cm/container_manager_windows.go:
```diff
@@ -111,16 +111,12 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-		cm.deviceManager, err = devicemanager.NewManagerImpl(nil, cm.topologyManager)
-		cm.topologyManager.AddHintProvider(cm.deviceManager)
-	} else {
-		cm.deviceManager, err = devicemanager.NewManagerStub()
```
Is ManagerStup used anywhere else? If not, should we remove it as well?

swatisehgal replied:
Thanks, doesn't look like it is being used anywhere else. Going to remove it.

klueska replied:
Its used in the e2e test

bart0sh replied:
git grep doesn't think so :)
```
ed@devel kubernetes (master) $ git grep NewManagerStub
pkg/kubelet/cm/container_manager_linux.go:              cm.deviceManager, err = devicemanager.NewManagerStub()
pkg/kubelet/cm/container_manager_windows.go:            cm.deviceManager, err = devicemanager.NewManagerStub()
pkg/kubelet/cm/devicemanager/manager_stub.go:// NewManagerStub creates a ManagerStub.
pkg/kubelet/cm/devicemanager/manager_stub.go:func NewManagerStub() (*ManagerStub, error) {
```

swatisehgal replied:
Yeah, I checked and we don't need the `NewManagerStub` for node e2e tests (in both [test/e2e_node/device_plugin_test.go ](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/device_plugin_test.go) and [test/e2e_node/device_manager_test.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/device_manager_test.go) and  ). My understanding is that for node e2e tests we run an instance of Kubelet and use device manager provided by Kubelet for testing and hence don't need a stub. Would you prefer if we left it in in case we need it in the future?

klueska replied:
This is what I was thinking of, which is different:
https://github.com/kubernetes/kubernetes/blob/master/test/images/sample-device-plugin/sampledeviceplugin.go#L89

bart0sh commented:
@swatisehgal there are still some golint findings we may want to fix before going GA:
```
ed@devel kubernetes (master) $ golint ./pkg/kubelet/cm/devicemanager/...
pkg/kubelet/cm/devicemanager/manager.go:168:1: exported method ManagerImpl.CleanupPluginDirectory should have comment or be unexported
pkg/kubelet/cm/devicemanager/manager.go:205:1: exported method ManagerImpl.PluginConnected should have comment or be unexported
pkg/kubelet/cm/devicemanager/manager.go:220:1: exported method ManagerImpl.PluginDisconnected should have comment or be unexported
pkg/kubelet/cm/devicemanager/manager.go:232:1: exported method ManagerImpl.PluginListAndWatchReceiver should have comment or be unexported
pkg/kubelet/cm/devicemanager/pod_devices.go:346:11: range var devId should be devID
pkg/kubelet/cm/devicemanager/pod_devices.go:375:1: exported function NewResourceDeviceInstances should have comment or be unexported
pkg/kubelet/cm/devicemanager/pod_devices.go:379:1: exported method ResourceDeviceInstances.Clone should have comment or be unexported
pkg/kubelet/cm/devicemanager/pod_devices_test.go:45:6: range var devId should be devID
pkg/kubelet/cm/devicemanager/checkpoint/checkpoint.go:80:1: exported function NewV2 should have comment or be unexported
pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:31:1: comment on exported type PodDevicesEntryV1 should be of the form "PodDevicesEntryV1 ..." (with optional leading article)
pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:66:1: comment on exported type DataV1 should be of the form "DataV1 ..." (with optional leading article)
pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:72:1: comment on exported function NewV1 should be of the form "NewV1 ..."
pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:93:1: comment on exported method DataV1.UnmarshalCheckpoint should be of the form "UnmarshalCheckpoint ..."
pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go:23:6: exported type RegistrationHandler should have comment or be unexported
pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go:27:6: exported type ClientHandler should have comment or be unexported
pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:33:6: exported type DevicePlugin should have comment or be unexported
pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:39:6: exported type Client should have comment or be unexported
pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:54:1: exported function NewPluginClient should have comment or be unexported
pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:108:18: method Api should be API
pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:39:6: exported type Server should have comment or be unexported
pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:57:1: exported function NewServer should have comment or be unexported
```

swatisehgal commented:
> @swatisehgal there are still some golint findings we may want to fix before going GA:
> 
> ```
> ed@devel kubernetes (master) $ golint ./pkg/kubelet/cm/devicemanager/...
> pkg/kubelet/cm/devicemanager/manager.go:168:1: exported method ManagerImpl.CleanupPluginDirectory should have comment or be unexported
> pkg/kubelet/cm/devicemanager/manager.go:205:1: exported method ManagerImpl.PluginConnected should have comment or be unexported
> pkg/kubelet/cm/devicemanager/manager.go:220:1: exported method ManagerImpl.PluginDisconnected should have comment or be unexported
> pkg/kubelet/cm/devicemanager/manager.go:232:1: exported method ManagerImpl.PluginListAndWatchReceiver should have comment or be unexported
> pkg/kubelet/cm/devicemanager/pod_devices.go:346:11: range var devId should be devID
> pkg/kubelet/cm/devicemanager/pod_devices.go:375:1: exported function NewResourceDeviceInstances should have comment or be unexported
> pkg/kubelet/cm/devicemanager/pod_devices.go:379:1: exported method ResourceDeviceInstances.Clone should have comment or be unexported
> pkg/kubelet/cm/devicemanager/pod_devices_test.go:45:6: range var devId should be devID
> pkg/kubelet/cm/devicemanager/checkpoint/checkpoint.go:80:1: exported function NewV2 should have comment or be unexported
> pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:31:1: comment on exported type PodDevicesEntryV1 should be of the form "PodDevicesEntryV1 ..." (with optional leading article)
> pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:66:1: comment on exported type DataV1 should be of the form "DataV1 ..." (with optional leading article)
> pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:72:1: comment on exported function NewV1 should be of the form "NewV1 ..."
> pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:93:1: comment on exported method DataV1.UnmarshalCheckpoint should be of the form "UnmarshalCheckpoint ..."
> pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go:23:6: exported type RegistrationHandler should have comment or be unexported
> pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go:27:6: exported type ClientHandler should have comment or be unexported
> pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:33:6: exported type DevicePlugin should have comment or be unexported
> pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:39:6: exported type Client should have comment or be unexported
> pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:54:1: exported function NewPluginClient should have comment or be unexported
> pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:108:18: method Api should be API
> pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:39:6: exported type Server should have comment or be unexported
> pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:57:1: exported function NewServer should have comment or be unexported
> ```

Sure, working to address this.

swatisehgal committed:
node: devicemgr: Address warnings from golint

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/test pull-kubernetes-e2e-gce-alpha-features

bart0sh commented:
/lgtm

bart0sh commented:
/retest

swatisehgal commented:
/test pull-kubernetes-e2e-gce-alpha-features

swatisehgal commented:
/test pull-kubernetes-e2e-gce-alpha-features

bart0sh commented:
@swatisehgal Failing test looks flaky in the test grid: https://k8s-testgrid.appspot.com/presubmits-kubernetes-nonblocking#pull-kubernetes-e2e-gce-alpha-features
Its failure doesn't look like it's caused by this PR, so we should keep re-testing I guess.

bart0sh commented:
/retest

swatisehgal commented:
/test pull-kubernetes-e2e-gce-alpha-features

klueska commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:
```diff
@@ -30,12 +30,14 @@ import (
 	SocketPath() string
 }
 
+// Client interface provides methods for accessing Device Plugin resources, API and unix socket.
```
This comment is incorrect

swatisehgal replied:
Updated, Thanks.

klueska commented on pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:
```diff
@@ -70,6 +74,7 @@ func (c *client) Connect() error {
 	return c.handler.PluginConnected(c.resource, c)
 }
 
+// Disconnect is for closing disconnecting client.
```
This comment is incorrect

swatisehgal replied:
Updated, Thanks.

klueska reviewed (commented):
Mostly minor comments. Looks good otherwise.

klueska commented:
/lgtm
/approve

swatisehgal commented:
/assign wojtek-t @marseel 
for SIG-scalability review as files in `cluster/gce` directory were updated.

/assign @dims @derekwaynecarr 
as `pkg/features/kube_features.go` was updated.

klueska commented:
also @endocrimes for `features`

dims commented on cluster/gce/config-default.sh:
```diff
@@ -255,15 +255,7 @@ if [[ (( "${KUBE_FEATURE_GATES:-}" == *"AllAlpha=true"* ) || ( "${KUBE_FEATURE_G
   RUN_CONTROLLERS="${RUN_CONTROLLERS:-*,endpointslice}"
 fi
 
-# Optional: set feature gates
```
i think we should leave this in!

dims replied:
`FEATURE_GATES="${KUBE_FEATURE_GATES:-}"`

swatisehgal replied:
I kept it in initially but `pull-kubernetes-verify` job failed because `FEATURE_GATES` was unused and therefore had to remove it. Here is the error message: 
```
In ./cluster/gce/config-test.sh line 160:
FEATURE_GATES=${KUBE_FEATURE_GATES:-}
^-----------^ SC2034: FEATURE_GATES appears unused. Verify use (or export if used externally).
In ./cluster/gce/config-default.sh line 259:
FEATURE_GATES="${KUBE_FEATURE_GATES:-}"
^-----------^ SC2034: FEATURE_GATES appears unused. Verify use (or export if used externally).
For more information:
  https://www.shellcheck.net/wiki/SC2034 -- FEATURE_GATES appears unused.
```
This is also captured in the comment above: https://github.com/kubernetes/kubernetes/pull/112980#issuecomment-1289968193.

dims replied:
@swatisehgal it's used in `config-helper.sh`.

please disable shell check for this line. see examples here:
https://cs.k8s.io/?q=SC2034&i=nope&files=&excludeFiles=&repos=

swatisehgal replied:
Fixed, Thanks.

dims commented on cluster/gce/config-default.sh:
```diff
@@ -255,15 +255,7 @@ if [[ (( "${KUBE_FEATURE_GATES:-}" == *"AllAlpha=true"* ) || ( "${KUBE_FEATURE_G
-FEATURE_GATES="${KUBE_FEATURE_GATES:-}"
-
 if [[ -n "${NODE_ACCELERATORS}" ]]; then
-    if [[ -z "${FEATURE_GATES:-}" ]]; then
```
this is OK to delete 262-266

dims commented on cluster/gce/config-test.sh:
```diff
@@ -156,9 +156,6 @@ if [[ (( "${KUBE_FEATURE_GATES:-}" = *"AllAlpha=true"* ) || ( "${KUBE_FEATURE_GA
 fi
 
-# Optional: set feature gates
-FEATURE_GATES=${KUBE_FEATURE_GATES:-}
```
same as above!

swatisehgal replied:
Please refer to the explanation on why this had to be removed [here](https://github.com/kubernetes/kubernetes/pull/112980#discussion_r1007828904).

swatisehgal commented:
/test pull-kubernetes-e2e-capz-windows-containerd
/test pull-kubernetes-e2e-gce-alpha-features

swatisehgal commented:
/retest

bart0sh commented:
/restest

bart0sh commented:
@swatisehgal could it be that alpha-features test fails because device manager tries to call device plugin, but no plugin is available?

swatisehgal commented:
> @swatisehgal could it be that alpha-features test fails because device manager tries to call device plugin, but no plugin is available?

I won't say I am 100% sure but it doesn't look like that is the case here. On inspection of the test args of pull-kubernetes-e2e-gce-alpha-features job [here](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L362) device manager e2e tests are not being executed as part of the job. Also the [build logs](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/112980/pull-kubernetes-e2e-gce-alpha-features/1585973179594051584/build-log.txt) show that the failures have nothing to do with device manager.

Both the jobs ([pull-kubernetes-e2e-gce-alpha-features](https://testgrid.k8s.io/presubmits-kubernetes-nonblocking#pull-kubernetes-e2e-gce-alpha-features) and [pull-kubernetes-e2e-capz-windows-containerd](https://testgrid.k8s.io/sig-windows-presubmit#pull-kubernetes-e2e-capz-windows-containerd) ) appear to be very flaky on testgrid. Also, both are not showing up as `required` so I assume that they wouldn't block PR merging once we have all the needed approvals. Will retest again anyway.

/retest

dims commented:
/retest
/approve
/lgtm

swatisehgal commented:
/test pull-kubernetes-integration

swatisehgal commented:
/test pull-kubernetes-e2e-kind

swatisehgal commented:
/retest
